src/regions/RichTextRegion.js

```diff
@@ -112,42 +114,93 @@ const Model = types
-      if (self.isText) return;
+    /**
+     * Main method to detect HTML range and its offsets for LSF region
+     * globalOffsets are used for:
+     * - internal use (get ranges to highlight quickly)
+     * - end users convenience
+     * - for emergencies (xpath invalid)
+     */
+    initRangeAndOffsets() {
+      const root = self._getRootNode();
+      let range;
 
-      const range = self._getRange(true);
+      // 0. Text regions are simple — just get range by offsets
+      if (self.isText) {
+        const { startOffset: start, endOffset: end } = self;
 
-      if (range && self.globalOffsets) {
-        const root = self._getRootNode(true);
+        self.globalOffsets = { start, end };
+        self.cachedRange = findRangeNative(start, end, root);
+        return;
+      }
+
+      // 1. first try to find range by xpath in original document
+      range = self._getRange({ useOriginalContent: true });
 
-        const rangeFromGlobal = findRangeNative(
-          self.globalOffsets.start,
-          self.globalOffsets.end,
-          root,
-        );
+      if (range) {
+        // we need this range in the visible document, so find it by global offsets
+        const originalRoot = self._getRootNode(true);
+        const [start, end] = rangeToGlobalOffset(range, originalRoot);
 
-        if (!rangeFromGlobal) return;
+        self.globalOffsets = { start, end };
+        self.cachedRange = findRangeNative(start, end, root);
+
+        return;
+      }
 
-        const normedRange = xpath.fromRange(rangeFromGlobal, root);
+      // 2. then try to find range on visible document
+      // that's for old buggy annotations created over dirty document state
+      range = self._getRange({ useOriginalContent: false });
 
-        if (!isDefined(normedRange)) return;
+      if (range) {
+        const [start, end] = rangeToGlobalOffset(range, root);
 
-        self.start = normedRange.start ?? self.start;
-        self.end = normedRange.end ?? self.end;
-        self.startOffset = normedRange.startOffset ?? self.startOffset;
-        self.endOffset = normedRange.endOffset ?? self.endOffset;
+        self.globalOffsets = { start, end };
+        self.cachedRange = range;
+
+        return;
+      }
+
+      // 3. if xpaths are broken use globalOffsets if given
+      if (self.globalOffsets && isDefined(root)) {
+        const { start, end } = self.globalOffsets;
+
+        self.globalOffsets = { start, end };
+        self.cachedRange = findRangeNative(start, end, root);
+
+        if (self.cachedRange) self._fixXPaths();
+
+        return;
       }
+
+      // 4. out of options — region is broken
+      // @todo show error in console and regions list
+      return undefined;
+    },
+
+    // fix XPaths when we found region by globalOffsets
+    _fixXPaths(range, root) {
+      const normedRange = xpath.fromRange(range, root);
+
+      if (!isDefined(normedRange)) return;
+
+      self.start = normedRange.start ?? self.start;
```

Comment:
why do we need this fallback to current values?
when it's possible that range is available, but some param of it is missing?